Feature Flag for Nested Mutations: CLI changes

severussundar opened the pull request:
## Why make this change?

-  Closes #1950 
    - Introduces a feature flag in the config for nested mutation operations.
    - Feature Flag format:
 
```json
     "runtime":{
      ...
       "graphql": {
           ...
           "nested-mutations": {
               "create": {
                     "enabled": true/false
                }
           }
       } 
     }
```
  - CLI Option: `--graphql.nested-create.enabled`. This option can be used along with `init` command to enable/disable nested insert operation.
 - By default, the nested mutation operations will be **disabled**.
 - Nested Mutation operations are applicable only for MsSQL database type. So, when the option `--graphql.nested-create.enabled` is used along with other database types, it is not honored and nested mutation operations will be disabled. Nested Mutation section will not be written to the config file. In addition, a warning will be logged to let users know that the option is inapplicable.  

## What is this change?

- `dab.draft.schema.json` - The schema file is updated to contain details about the new fields
- `InitOptions` - A new option `--graphql.nested-create.enabled` is introduced for the `init` command.
- `NestedCreateOptionsConverter` - Custom converter to read & write the options for nested insert operation from/to the config file respectively.
- `NestedMutationOptionsConverter` - Custom converter to read & write the options for nested mutation operations from/to the config file respectively.
- `GraphQLRuntimeOptionsConverterFactory` - Updates the logic for reading and writing the graphQL runtime section of the config file. Incorporates logic for reading and writing the nested mutation operation options.
- `dab-config.*.json`/`Multidab-config.*.json` - All the reference config files are updated to include the new Nested Mutation options

## How was this tested?

- [x] Integration Tests
- [x] Unit Tests
- [x] Manual Tests

## Sample Commands

- **Nested Create Operations are enabled**:
`dab init --database-type mssql --connection-string connString --graphql.nested-create.enabled true`
![image](https://github.com/Azure/data-api-builder/assets/11196553/c1821897-1553-46d7-97d2-bf31b7f6178d)



- **Nested Create Operations are disabled**:
`dab init --database-type mssql --connection-string connString --graphql.nested-create.enabled false`
![image](https://github.com/Azure/data-api-builder/assets/11196553/ea421080-beb8-4f01-a2c9-99916b8b83cc)



- **When --graphql.nested-create.graphql option is not used in the init command**:
`dab init --database-type mssql --connection-string connString`
![image](https://github.com/Azure/data-api-builder/assets/11196553/d6f1d56c-a553-4dbf-8ad1-e813edc4274d)



- **When --graphql.nested-create.graphql option is used with a database type other than MsSQL**:
![image](https://github.com/Azure/data-api-builder/assets/11196553/f9cdda69-f0bd-4e9d-8f65-dd1f0df48402)

severussundar commented:
/azp run

seantleonard commented on src/Config/RuntimeConfigLoader.cs:
```diff
@@ -170,7 +170,8 @@ public RuntimeConfigLoader(string? connectionString = null)
         options.Converters.Add(new EntityRestOptionsConverterFactory(replaceEnvVar));
         options.Converters.Add(new EntityActionConverterFactory());
         options.Converters.Add(new DataSourceFilesConverter());
-        options.Converters.Add(new EntityCacheOptionsConverterFactory());
```
This removal causes unit test errors. need to add this back.

severussundar replied:
Thanks for spotting this. 

Removed this accidentally. Added back now.

abhishekkumams commented:
/azp run

abhishekkumams reviewed (approved):
LGTM

Aniruddh25 commented on schemas/dab.draft.schema.json:
```diff
@@ -175,6 +175,23 @@
+              "type": "object",
+              "description": "Feature flags for nested mutation operations",
+              "properties": {
+                "inserts":{
```
nit: we should call this as `create` since that aligns with our existing nomenclature for actions, permission etc

severussundar replied:
Renamed `insert` to `create`

Aniruddh25 commented on schemas/dab.draft.schema.json:
```diff
@@ -175,6 +175,23 @@
+              "properties": {
+                "inserts":{
+                  "type": "object",
+                  "description": "Options for nested insert operations",
```
nit: nested create, please make the change from insert to create wherever applicable.

severussundar replied:
I've updated all the relevant places

Aniruddh25 commented on src/Cli/ConfigGenerator.cs:
```diff
@@ -114,6 +114,29 @@ public static bool TryCreateRuntimeConfig(InitOptions options, FileSystemRuntime
 
+            bool isNestedInsertEnabledForGraphQL;
+
+            // Nested mutation operations are applicable only for MSSQL database. When the option --graphql.nested-insert.enabled is specified for other database types,
```
nit: Please create a tracking issue to add support for other Sql types for nested create, and document it here

severussundar replied:
I've created the issue : [#2001](https://github.com/Azure/data-api-builder/issues/2001) to track. Added a comment with the issue.

Aniruddh25 commented on src/Cli/Commands/InitOptions.cs:
```diff
@@ -36,7 +36,8 @@ public class InitOptions : Options
             CliBool graphqlEnabled = CliBool.None,
-            CliBoolean restRequestBodyStrict = CliBoolean.None,
+            CliBool restRequestBodyStrict = CliBool.None,
+            CliBool nestedInsertOperationEnabled = CliBool.None,
```
To track as a separate issue, I think we should add a different `dab` command that helps to enable or disable a feature, should not be only possible through the `init` command - which is the case today.

severussundar replied:
I've created issue [#2000 ](https://github.com/Azure/data-api-builder/issues/2000) for tracking this.

Aniruddh25 commented on src/Config/ObjectModel/GraphQLRuntimeOptions.cs:
```diff
@@ -3,7 +3,7 @@
 namespace Azure.DataApiBuilder.Config.ObjectModel;
 
-public record GraphQLRuntimeOptions(bool Enabled = true, string Path = GraphQLRuntimeOptions.DEFAULT_PATH, bool AllowIntrospection = true)
+public record GraphQLRuntimeOptions(bool Enabled = true, string Path = GraphQLRuntimeOptions.DEFAULT_PATH, bool AllowIntrospection = true, NestedMutationOptions? NestedMutationOptions = null)
```
nit: please split into multiple lines - the constructor parameters have become far too many

Aniruddh25 commented on src/Service.Tests/dab-config.PostgreSql.json:
```diff
@@ -14,7 +14,12 @@
       "path": "/graphql",
-      "allow-introspection": true
+      "allow-introspection": true,
+      "nested-mutations": {
```
Doesnt the old config work as is? Why do we need to explicitly set this to false?

severussundar replied:
The old config (without the `nested-mutations` section ) works correctly. Added the section in the reference config files to showcase how the section would look like

Aniruddh25 commented on src/Service.GraphQLBuilder/Azure.DataApiBuilder.Service.GraphQLBuilder.csproj:
```diff
@@ -6,6 +6,7 @@
     <Nullable>enable</Nullable>
     <OutputPath>$(BaseOutputPath)\engine</OutputPath>
     <EmbedUntrackedSources>true</EmbedUntrackedSources>
+    <NoWarn>NU1603</NoWarn>
```
nit: If we are adding this property to all projects, is it better to maintain the property group at root level similar to the `Directory.Packages.props`?

severussundar replied:
This change is not inteded to be added as part of this PR. This is showing up in the changeset because the target branch was not  updated with the latest main whereas this branch is. 

I've updated the target branch with the latest changes from main. So, this change shouldn't come up anymore.

Aniruddh25 commented on src/Cli.Tests/EndToEndTests.cs:
```diff
@@ -131,6 +131,44 @@ public void TestInitializingRestAndGraphQLGlobalSettings()
+    /// <param name="isNestedInsertsEnabled">Value for the nested insert enabled flag in the init command.</param>
+    /// <param name="expectedValueForNestedInsertEnabledFlag"> Expected value for the nested insert enabled flag in the config file.</param>
+    [DataTestMethod]
+    [DataRow(CliBool.True, true, DisplayName = "Nested Insert operation is enabled the config file by specifying '--graphql.nested-insert.enabled true'")]
```
```suggestion
    [DataRow(CliBool.True, true, DisplayName = "Nested Insert operation is enabled by specifying '--graphql.nested-insert.enabled true'")]
```

Aniruddh25 commented on src/Cli.Tests/EndToEndTests.cs:
```diff
@@ -131,6 +131,44 @@ public void TestInitializingRestAndGraphQLGlobalSettings()
+    /// <param name="expectedValueForNestedInsertEnabledFlag"> Expected value for the nested insert enabled flag in the config file.</param>
+    [DataTestMethod]
+    [DataRow(CliBool.True, true, DisplayName = "Nested Insert operation is enabled the config file by specifying '--graphql.nested-insert.enabled true'")]
+    [DataRow(CliBool.False, false, DisplayName = "Nested Insert operation is disabled the config file by specifying '--graphql.nested-insert.enabled false'")]
```
```suggestion
    [DataRow(CliBool.False, false, DisplayName = "Nested Insert operation is disabled by specifying '--graphql.nested-insert.enabled false'")]
```

Aniruddh25 commented on src/Cli.Tests/EndToEndTests.cs:
```diff
@@ -131,6 +131,44 @@ public void TestInitializingRestAndGraphQLGlobalSettings()
+        Assert.AreEqual(DatabaseType.MSSQL, runtimeConfig.DataSource.DatabaseType);
+        Assert.IsNotNull(runtimeConfig.Runtime);
+        Assert.IsNotNull(runtimeConfig.Runtime.GraphQL);
+        Assert.IsNotNull(runtimeConfig.Runtime.GraphQL.NestedMutationOptions);
```
in case the option is not specified i.e. `isNestedInsertedEnabled` is null, there should be no change to the config. We should not be writing out the nested-mutations section at all. The config size will not increase as a result when people upgrade

severussundar replied:
Writing out all the fields has been the approach we've adopted for other fields, and this is in line with that.
Some examples of this include: 
1. Writing out the `path` and `request-body-strict` fields even if `rest` is disabled
2. Likewise, `path`, `allow-introspection` fields for `graphQL`
3.  In the entities section, we again write out fields like `path`, `singular`, `plural` names, `source.fields`, `target.fields` of relationship (when fields are not specified),  `rest.methods`, `graphql.operations` and more. 

Also, writing out the fields has the following advantages,
1. Users get to know the presence of this new section in the config file. They might not always closely follow the latest changes to the documentation. 
2. Avoids any ambiguity on the default behavior when the option is not used. Even though, we would document the default behavior, they might not be caught with the latest additions to the docs.


I've tested to the changes with an old config file (without this new section) that it deserializes correctly and REST, GraphQL API requests continue to work fine. So, this let's be backwards compatible.

seantleonard replied:
are there tests for this: "I've tested the changes with an old config file (without the nested mutation fields) that it deserializes correctly and REST, GraphQL API requests work fine." 
When someone doesn't use the CLI, they won't have the new fields output to their config file. Need to test that excluding the new fields from the config or setting them to null has no negative effects.

severussundar replied:
I noticed that we don't have any test that validates the backward compatibility aspect, specifically an older version of config file working successfully with its subsequent engine versions.  This cab be a broader discussion on how (and if) we want to have such tests.

To validate this particular change, I've added a sanity test - This test uses hard coded config Json strings to mimic the scenario where users hand-edit the config file instead of using CLI. This config file is used to start the DAB engine and REST and GraphQL requests are executed against it and validated.

severussundar replied:
After discussion in the scrum on 2nd Feb 2024: Need to modify the logic to not write the fields with default values when the option is not used in the command. The older approach has been to write out all the fields, but to avoid bloat in the config file, going forward, we will not write out the fields (ones corresponding to options that are not used) to the config file. This also avoids any issues that could arise when merging two config files.

seantleonard replied:
Making the above mentioned update will reduce the size of this pr too hopefully  since no longer updating all the verify.txt files.

Aniruddh25 replied:
Discussed in scrum the importance of not writing out to file the default behavior if the property is null

severussundar replied:
I've updated the logic to not write out fields with default values when the option is not used in the command

Aniruddh25 reviewed (changes requested):
A comprehensive change fixing all the aspects where this option would touch, but requesting changes for some assumptions:

1. insert => create
2. nested mutation section should not be present in config if its not specified in init

severussundar commented:
/azp run

severussundar commented:
/azp run

severussundar commented:
/azp run

severussundar commented:
/azp run

seantleonard commented on src/Cli.Tests/InitTests.cs:
```diff
@@ -409,6 +409,73 @@ public Task GraphQLPathWithoutStartingSlashWillHaveItAdded()
         }
 
+        /// <summary>
+        /// Test to validate the config is correctly generated with different database types and various options for --graphql.nested-create.enabled flag.
```
Can you elaborate on what you're verifying? contents of generated config? are you also validating behavior when dab get various db types other than mssql?

severussundar replied:
I'm validating the contents of the generated config file. I've updated the description

seantleonard commented on src/Cli/ConfigGenerator.cs:
```diff
@@ -286,6 +311,30 @@ private static bool TryDetermineIfApiIsEnabled(bool apiDisabledOptionValue, CliB
+        {
+            if (nestedCreateEnabledOptionValue is CliBool.None)
+            {
+                isNestedCreateEnabledForGraphQL = false;
```
still confused here. Set `isNestedCreateEnabledForGraphQL` to false but return true for `IsNestedCreateOperationEnabled`

severussundar replied:
The return value indicates whether the function ran into issues when determining nested create is enabled or not for graphQL. The reason for doing this is to account for any errors that might arise when parsing the value entered to a boolean value.

But this is unnecessary, because the function takes a CliBool type parameter which can take only 3 values - None, True or False. And, this function is called only when the value is either True/False. So, there should not be any such errors when doing `bool.parse(...)`

So, I've updated the function to directly return the value of whether nested create is enabled or not.

seantleonard commented on src/Service.Tests/Configuration/ConfigurationTests.cs:
```diff
@@ -1579,6 +1579,115 @@ public void TestConnectionStringEnvVarHasHighestPrecedence()
+            // The config file is constructed by merging the hard-coded json strings to mimic the scenario where config file is
+            // hand-edited (instead of using CLI) by the users.
+            string configJson = TestHelper.AddPropertiesToJson(TestHelper.BASE_CONFIG, entityJson);
+            RuntimeConfigLoader.TryParseConfig(configJson, out RuntimeConfig deserializedConfig, logger: null, GetConnectionStringFromEnvironmentConfig(environment: TestCategory.MSSQL));
```
tests which validate the results of `RuntimeConfigLoader.TryParseConfig(...)` would be helpful to know how the new nested insert config behaves when unexpected values are assigned or not assigned:
- nulls assigned to properties, 
```json
{
  "nested-mutations": null
}
```
- empty objects 
```json
{
  "nested-mutations": {
    "create": {}
  }
}

{
  "nested-mutations": {}
}
```

severussundar replied:
Ack. Looking into this.

severussundar replied:
I've added tests to validate these scenarios: `ConfigurationTests.ValidateDeserializationOfConfigWithNullOrEmptyInvalidNestedMutationSection()`

seantleonard reviewed (commented):
Pending some open threads to resolve/respond

severussundar commented on src/Cli/ConfigGenerator.cs:
```diff
@@ -286,6 +311,30 @@ private static bool TryDetermineIfApiIsEnabled(bool apiDisabledOptionValue, CliB
+        {
+            if (nestedCreateEnabledOptionValue is CliBool.None)
+            {
+                isNestedCreateEnabledForGraphQL = false;
+                return true;
```
The return value is used to indicate any problems encountered when determining whether nested create is enabled or not. In this code block, when the flow reaches L324, there would be no errors hit, so `true` is returned.

True/False indicating whether nested create is enabled or not, is stored in the out variable `isNestedCreateEnabledForGraphQL`

severussundar replied:
I've explained this in the other comment

severussundar commented on src/Service.Tests/Caching/CachingConfigProcessingTests.cs:
```diff
@@ -416,7 +416,12 @@ private static string GetRawConfigJson(string globalCacheConfig, string entityCa
+                        ""create"": {
+                            ""enabled"": false
+                        }
+                    }
```
After updating the logic to not write out the fields when the option is not used, the expected json strings do not need to have these fields. Removed this.

severussundar commented on src/Service.Tests/Multidab-config.MsSql.json:
```diff
@@ -15,7 +15,12 @@
+        "create": {
+          "enabled": false
+        }
+      }
```
I've updated the value to `true`.

severussundar commented:
/azp run

abhishekkumams commented on schemas/dab.draft.schema.json:
```diff
@@ -175,6 +175,23 @@
+              "type": "object",
+              "description": "Feature flags for nested mutation operations",
+              "properties": {
+                "create":{
```
in the pr description, i see `insert`, but here it's `create` ?

severussundar replied:
PR description is not update. Will update

severussundar replied:
Updated

abhishekkumams commented on src/Cli/ConfigGenerator.cs:
```diff
@@ -286,6 +307,21 @@ private static bool TryDetermineIfApiIsEnabled(bool apiDisabledOptionValue, CliB
+        /// <returns>True/False</returns>
+        private static bool IsNestedCreateOperationEnabled(CliBool nestedCreateEnabledOptionValue)
+        {
+            if (nestedCreateEnabledOptionValue is CliBool.None)
```
No need to parse. 
```
if (nestedCreateEnabledOptionValue is CliBool.None or CliBool.False){
     return false;
}

return true;
```

severussundar replied:
Incorporated

abhishekkumams commented:
how will the engine respond to the below scenarios (what will be the value default values)? Also add test to make sure in below cases parsing doesn't fail.

**Case 1:**
```
"nested-mutations": null
```

**Case 2:**
```
"nested-mutations": { }
```


**Case 3:**
```
"nested-mutations": {
    "create": null
}
```

**Case 4:**

```
"nested-mutations": {
    "create": {}
}
```

abhishekkumams commented on schemas/dab.draft.schema.json:
```diff
@@ -175,6 +175,23 @@
+              "properties": {
+                "create":{
+                  "type": "object",
+                  "description": "Options for nested create operations",
```
add "additionalProperties": false.  So, that if any property is there or if there is a typo, it is being caught by the schema validator

abhishekkumams commented on src/Service.Tests/Configuration/ConfigurationTests.cs:
```diff
@@ -1579,6 +1616,119 @@ public void TestConnectionStringEnvVarHasHighestPrecedence()
+        [TestCategory(TestCategory.MSSQL)]
+        public async Task SanityTestForRestAndGQLRequestsWithoutNestedMutationFeatureFlagSection()
+        {
+            // The config file is constructed by merging the hard-coded json strings to mimic the scenario where config file is
```
```suggestion
            // The configuration file is constructed by merging hard-coded JSON strings to simulate the scenario where users manually edit the configuration file (instead of using the CLI).
```

ayush3797 commented:
QQ @severussundar: Is there validation which ensures that the feature is set to true only for `MsSql`?

ayush3797 commented:
> QQ @severussundar: Is there validation which ensures that the feature can be set to true only for `MsSql`?

Nvm, referred to the example in description.

ayush3797 commented on src/Config/ObjectModel/NestedInsertOptions.cs:
```diff
@@ -0,0 +1,21 @@
+/// Options for nested create operations.
+/// </summary>
+/// <param name="Enabled"> Indicates whether nested create operation is enabled.</param>
+public class NestedCreateOptions
```
Why not use a more generic name like `NestedMutationOptions`? Makes more sense when we start supporting nested updates as well?

severussundar commented:
> how will the engine respond to the below scenarios (what will be the value default values)? Also add test to make sure in below cases parsing doesn't fail.
> 
> **Case 1:**
> 
> ```
> "nested-mutations": null
> ```
> 
> **Case 2:**
> 
> ```
> "nested-mutations": { }
> ```
> 
> **Case 3:**
> 
> ```
> "nested-mutations": {
>     "create": null
> }
> ```
> 
> **Case 4:**
> 
> ```
> "nested-mutations": {
>     "create": {}
> }
> ```

I've added tests for all these 4 scenarios to validate that deserialization is successful. In all these cases, after deserialization `NestedMutationOptions` will have a value of `null`. I've added assert to validate this too.
property will be `null`

severussundar commented:
/azp run

severussundar commented:
/azp run

Aniruddh25 reviewed (commented):
Posting comments so far

Aniruddh25 commented on src/Cli.Tests/Snapshots/InitTests.VerifyCorrectConfigGenerationWithNestedMutationOptions_09cf40a5c545de68.verified.txt:
```diff
@@ -0,0 +1,34 @@
+    GraphQL: {
+      Enabled: true,
+      Path: /graphql,
+      AllowIntrospection: true
```
Do we get the snapshots for each DataRow being tested?

severussundar replied:
Yes

Aniruddh25 commented on src/Config/Converters/GraphQLRuntimeOptionsConverterFactory.cs:
```diff
@@ -35,11 +55,79 @@ private class GraphQLRuntimeOptionsConverter : JsonConverter<GraphQLRuntimeOptio
+                        break;
+                    }
+
+                    string? propertyName = reader.GetString();
```
What if propertyName is null here? Is it possible? If not, why is its type a nullable string?

severussundar replied:
The type is kept nullable because of the return type of `reader.GetString()`. `propertyName` is not null when the flow reaches this point. A `null` property name is invalid and should get caught before the flow reaches here. 

However, I've added a null check, and an exception is thrown if `propertyName` turns out to be `null`.

Aniruddh25 commented on src/Config/Converters/NestedMutationOptionsConverter.cs:
```diff
@@ -0,0 +1,88 @@
+                return nestedMutationOptions;
+            }
+
+            throw new JsonException();
```
Please provide a descriptive exception message here

severussundar replied:
Updated

Aniruddh25 commented on src/Config/ObjectModel/NestedMutationOptions.cs:
```diff
@@ -0,0 +1,31 @@
+public class NestedMutationOptions
+{
+    // Options for nested create operation.
+    [JsonPropertyName("insert")]
```
why is the json property name insert?

severussundar replied:
Good catch.

Actually, the `JsonPropertyName` is not needed since we have custom converters. Removed this.

Aniruddh25 commented on src/Config/ObjectModel/NestedInsertOptions.cs:
```diff
@@ -0,0 +1,21 @@
+/// </summary>
+/// <param name="Enabled"> Indicates whether nested create operation is enabled.</param>
+public class NestedCreateOptions
+{
```
File name still says NestedInsertOptions.cs. please rename

severussundar replied:
Renamed

Aniruddh25 commented on src/Config/Converters/NestedMutationOptionsConverter.cs:
```diff
@@ -0,0 +1,88 @@
+        /// <inheritdoc/>
+        public override NestedMutationOptions? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            if (reader.TokenType == JsonTokenType.Null)
```
To simplify usage, is it possible to have:

`"nested-mutations" : true` 
to imply any supported nested mutations are enabled. Could be a separate PR.

severussundar replied:
This would involve a slightly broader discussion and also requires a new option. 

The current option `--graphql.nested-create.enabled` at the granularity level of `create` operation. To enable/disable something that is at the level of all mutation operations, we will need a new option. 

Ideally, should be taken up in a new issue after discussion.

severussundar replied:
I've created this issue --->  [#2089 ](https://github.com/Azure/data-api-builder/issues/2089)

Aniruddh25 commented on src/Config/Converters/NestedCreateOptionsConverter.cs:
```diff
@@ -0,0 +1,72 @@
+                return nestedCreateOptions;
+            }
+
+            throw new JsonException();
```
Needs an useful exception message

severussundar replied:
Updated

Aniruddh25 commented on src/Config/Converters/NestedCreateOptionsConverter.cs:
```diff
@@ -0,0 +1,72 @@
+                        break;
+                    }
+
+                    string? propertyName = reader.GetString();
```
what if propertyName is null here?

severussundar replied:
Added a null check and an exception is thrown when `propertyName` turns out to be `null`

Aniruddh25 commented on src/Service.Tests/Configuration/ConfigurationTests.cs:
```diff
@@ -1581,6 +1618,119 @@ public void TestConnectionStringEnvVarHasHighestPrecedence()
+        ///     }
+        /// }
+        /// 
+        /// For all the above mentioned scenarios, the expected value for NestedMutationOptions field in null.
```
```suggestion
        /// For all the above mentioned scenarios, the expected value for NestedMutationOptions field is null.
```

Aniruddh25 reviewed (changes requested):
Still some references to `insert` instead of `create`

severussundar commented on src/Service.Tests/dab-config.MySql.json:
```diff
@@ -17,7 +17,7 @@
       "allow-introspection": true,
       "nested-mutations": {
         "create": {
-          "enabled": false
```
Good catch. I've removed this section in the config files for other DBs (apart from MsSQL)

severussundar commented on src/Service.Tests/TestHelper.cs:
```diff
@@ -194,6 +194,130 @@ public static RuntimeConfig AddMissingEntitiesToConfig(RuntimeConfig config, str
+              ""graphql"": {
+                ""path"": ""/graphql"",
+                ""allow-introspection"": true,
+                ""nested-mutations"": null
```
Added a common string to get rid of duplication.

michaelstaib commented:
Are you guys aware that nested mutations in GraphQL are not spec compliant?

severussundar commented:
/azp run

severussundar commented:
> Are you guys aware that nested mutations in GraphQL are not spec compliant? Its specifically called out by the specification text. And in the TSC there is consensus not to change it. There are discussions on namespacing to solve the actual issue that users have. Nested mutations in itself remove the guarantees for how mutations are executed and thus break all GraphQL executors.


Hi @michaelstaib, thank you very much for sharing this perspective.

Here's what we refer to as a nested mutation operation. Right now, we are working only towards Nested Create operation.

Nested Create refers to the ability to create items in the related entity (or entities) in addition to the top-level entity in a single mutation operation. 

For example,
Let's say we have the following entities defined: Book, Publisher, Reviews, Authors, BookDetails. 

Relationship among these entities:
1. Book: Publisher -  N:1
2. Book: Reviews - 1:N
3. Book: Authors- M:N
4. BookDetails - 1:1

For this setup, through a single nested create operation, users would be able to 
1. Create a book and along with it,
      a) Create a list of reviews for it 
      b) Create a publisher for it
      c) Create an author for it. 

The association between the created book item, publisher item, author items, review items would be handled by DAB.

From the user's perspective, they would write a graphQL query such as this

![image](https://github.com/Azure/data-api-builder/assets/11196553/9b36e5b8-ac0b-46fe-9d55-7ccb278364ab)

When I checked the graphQL specs, I did not find anything that disallows such a mutation request. 

Could you please kindly share,
1. If by `nested mutations`, you were referring to a different operation or the same as this one.
2. The links to the spec that disallows nested mutation operation. It would be very helpful for us to read through and double check.

Aniruddh25 commented on src/Service.Tests/TestHelper.cs:
```diff
@@ -220,103 +232,56 @@ public static RuntimeConfig AddMissingEntitiesToConfig(RuntimeConfig config, str
-          "{" +
-            SAMPLE_SCHEMA_DATA_SOURCE + "," +
+
+            RUNTIME_REST_GRAPQL_OPTIONS_CONFIG_JSON +
```
```suggestion
            RUNTIME_REST_GRAPHQL_OPTIONS_CONFIG_JSON +
```
this should be fixed in all its occurrences

Aniruddh25 reviewed (approved):
Need to fix a typo, LGTM otherwise.

Aniruddh25 reviewed (dismissed):
Seems like this open issue in the graphql spec: https://github.com/graphql/graphql-spec/issues/252 contradicts with what we are calling as nested mutations. I think we should rename the feature to avoid confusion in its usage to something like `multiple-mutations` + `multiple-create`, since our feature specifically means performing multiple mutations on related entities in a serial manner.

Nested-mutations as per the open issue in the community refers to having a `mutation` operation under another `query`/`mutation` node - which is not what we plan to do here.

Thanks to @michaelstaib for bringing this up.

michaelstaib commented:
@severussundar what you are doing is correct and spec compliant. It looks very similar to what is referred to as batch mutation where the input represents not a single mutation but potentially multiple.

```graphql
mutation {
  updateCart(input: { productToAdd: [  ...], productsToRemove: [  ...] }) {   ...}
}
```

seantleonard commented on src/Config/Converters/NestedMutationOptionsConverter.cs:
```diff
@@ -0,0 +1,88 @@
+
+            writer.WriteStartObject();
+
+            if (value.NestedCreateOptions is not null)
+            {
+                _nestedCreateOptionsConverter.Write(writer, value.NestedCreateOptions, options);
+            }
+
```
if `(value.NestedCreateOptions` is null, then i'd argue that we wouldn't want to write the nested mutations property right? that would be invalid config to provide:
1.  `"nested-mutations": {}` 
2.  `"nested-mutations": null`

severussundar replied:
In L71, I am performing a `null` check against `value` which is of type `NestedMutationOptions`. If it is `null`, then, nothing will be written to the config file. So, the config file created using CLI will not have null or empty values for `nested-mutations` property. 


Right now, `NestedMutationOptions` will be non-null, only when `NestedCreateOptions` is also not null. 

However, in the future, if we introduce `NestedUpdateOptions`, then it is possible that `NestedMutationOptions` is non-null but `NestedCreateOptions` is null. But, that should also get taken care of as a `null` check is performed on L80 (before writing the fields related to nested create out to the config file).

severussundar replied:
I ack that configs such as `"nested-mutations": {}` and `"nested-mutations": null` are possible as a result of users manually editing the config file instead of using CLI. I've handled these scenarios when reading the config file and added tests to validate the same.

seantleonard commented on src/Service.Tests/Configuration/ConfigurationTests.cs:
```diff
@@ -1581,6 +1618,119 @@ public void TestConnectionStringEnvVarHasHighestPrecedence()
+        /// </summary>
+        /// <param name="baseConfig">Base Config Json string.</param>
+        [DataTestMethod]
+        [DataRow(TestHelper.BASE_CONFIG_NULL_NESTED_MUTATIONS_FIELD, DisplayName = "Validate successful deserialization when nested mutation section is null")]
```
more self descriptive displayname:
Instead of: "Validate successful deserialization",
i propose:
"NestedMutationOptions field deserialized as null when {test case}"

seantleonard commented on src/Service.Tests/Configuration/ConfigurationTests.cs:
```diff
@@ -1581,6 +1618,119 @@ public void TestConnectionStringEnvVarHasHighestPrecedence()
+            Assert.IsTrue(RuntimeConfigLoader.TryParseConfig(configJson, out RuntimeConfig deserializedConfig));
+            Assert.IsNotNull(deserializedConfig.Runtime);
+            Assert.IsNotNull(deserializedConfig.Runtime.GraphQL);
+            Assert.IsNull(deserializedConfig.Runtime.GraphQL.NestedMutationOptions);
```
is `{ "graphql" : { "nested-mutatations": null } }` valid config?

severussundar replied:
1.  
```json  
  "nested-mutations": null
```

2.
```json
  "nested-mutations": {}
```

3.
```json
  "nested-mutations":{
      "create": { }
  }
```
4.
```json
  "nested-mutations":{
      "create": null 
  }
```


For the above four possibilities, I've taken the following approach:
1. Deserialization of the config file will succeed - there will be no errors (engine would not crash).
2. After deserialization, the `NestedMutationOptions` will be `null`. So, this will result in nested mutation operations being disabled.

seantleonard reviewed (commented):
Some questions about conditions and valid config representations of nested mutations section. Should be good to go after addressing.

severussundar commented on src/Cli.Tests/EndToEndTests.cs:
```diff
@@ -131,6 +131,69 @@ public void TestInitializingRestAndGraphQLGlobalSettings()
+            args.Add(isNestedCreateEnabled.ToString()!);
+        }
+
+        Program.Execute(args.ToArray(), _cliLogger!, _fileSystem!, _runtimeConfigLoader!);
+
+        Assert.IsTrue(_runtimeConfigLoader!.TryLoadConfig(
+            TEST_RUNTIME_CONFIG_FILE,
+            out RuntimeConfig? runtimeConfig,
+            replaceEnvVar: true));
```
`Program.Execute()`. 
It is called with the various possible combinations of arguments array as applicable for each test scenario. (Running a CLI command invokes the same function with the applicable arguments). Calling this function results in the creation of the config file.

 `_runtimeConfigLoader!.TryLoadConfig(...)` is used for validation. The generated config file is deserialized into a `RuntimeConfig` type object and the fields are validated if they contain the expected values.

severussundar commented on src/Config/ObjectModel/NestedCreateOptions.cs:
```diff
@@ -0,0 +1,21 @@
+    /// <summary>
+    /// Indicates whether nested create operation is enabled.
+    /// </summary>
+    public bool Enabled;
```
JsonProperty declaration is not needed. The serialization and deserialization of the property is taken care by the custom converters.

severussundar commented on src/Config/RuntimeConfigLoader.cs:
```diff
@@ -164,13 +164,15 @@ public RuntimeConfigLoader(string? connectionString = null)
         options.Converters.Add(new EnumMemberJsonEnumConverterFactory());
         options.Converters.Add(new RestRuntimeOptionsConverterFactory());
-        options.Converters.Add(new GraphQLRuntimeOptionsConverterFactory());
+        options.Converters.Add(new GraphQLRuntimeOptionsConverterFactory(replaceEnvVar));
```
Before the introduction of nested/multiple mutation related fields in `GraphQLRuntimeOptions`, the fields present were `path`, `enabled` and `allow-introspection` - all of them are either boolean or string type. 

So, in `GraphQLRuntimeOptionsConverterFactory`, it was sufficient to have the deserialization logic as 
`JsonSerializer.Deserialize<GraphQLRuntimeOptions>(ref reader, innerOptions)`
 
But, with the `nested-mutations` field is of object type. In addition, it contains another object type field `create`. So, these fields require a different deserialization logic from that of fields such as `path`, `enabled`, etc. 

When introducing separate deserialization logic, I noticed that we support env variable replacement for `path` field. To incorporate this in the new logic, `replaceEnvVar` property was needed.

severussundar commented on src/Service.Tests/Multidab-config.MsSql.json:
```diff
@@ -15,7 +15,12 @@
+      "allow-introspection": true,
+      "nested-mutations": {
+        "create": {
+          "enabled": true
```
Considering these files serve as a reference to the users - giving them an idea of how the resulting config will look like in the latest version, adding either `true` or `false` should be okay.

For integration tests, the config file that gets generated using the commands present in `*-commands.txt` gets utilized. But your question is valid and is one of the things I'm thinking through for adding tests in my other PR - essentially, how do we run integration tests with nested mutation operations as enabled/disabled. The challenge is that the config file gets generated once (and is utilized by all the integration tests) in the pipeline. 

One approach I'm considering - For tests specifically wanting to validate something when nested mutation is enabled/disabled, during the test setup phase, the config file can be read, and the `enabled` field can be edited to `true`/`false` as necessary and the file can be written back. Likewise, in the cleanup phase, the `enabled` field can be reset.

seantleonard replied:
Tests that rely on the value of `nestedmutations.enabled==true/false` should get their own custom config file created during test startup and not rely on generic test files used in generic mega classes like sqltestbase. Examples of this can be found in ConfigurationTests.cs

severussundar replied:
Yeah, I agree.

severussundar commented on src/Cli.Tests/InitTests.cs:
```diff
@@ -409,6 +409,73 @@ public Task GraphQLPathWithoutStartingSlashWillHaveItAdded()
         }
 
+        /// <summary>
+        /// Test to validate the contents of the config file generated when init command is used with --graphql.nested-create.enabled flag option for different database types.
```
Updated the test description.

severussundar commented on src/Cli/ConfigGenerator.cs:
```diff
@@ -285,6 +306,21 @@ private static bool TryDetermineIfApiIsEnabled(bool apiDisabledOptionValue, CliB
+            }
+
+            return true;
+        }
```
Good catch. Updated.

severussundar commented on src/Cli.Tests/EndToEndTests.cs:
```diff
@@ -131,6 +131,69 @@ public void TestInitializingRestAndGraphQLGlobalSettings()
     }
 
+    /// <summary>
+    /// Test to validate the successful generation of config file with the --graphql.nested-create.enabled option of the init command.
```
This test validates the usage of `--graphql.nested-create.enabled` option for all possible combinations: 

a) For DB types than MsSQL - irrespective of whether the option is used or not, nothing gets written to the config file as the option is inapplicable. 
b) For MsSQL: 
       i) When the option is not used (CLI interprets this as `CliBool.None`), nothing gets written out to the config file.
       ii) When true/false is used, that exact value is written to the config file.
              
The config file is created using these different options; then it is later deserialized and the properties are validated against the expected value (as applicable for that scenario).

I've updated the description with these details.

severussundar commented on src/Cli.Tests/EndToEndTests.cs:
```diff
@@ -131,6 +131,69 @@ public void TestInitializingRestAndGraphQLGlobalSettings()
+    [DataTestMethod]
+    [DataRow(CliBool.True, "mssql", DatabaseType.MSSQL, DisplayName = "Init command with '--graphql.nested-create.enabled true' for MsSql database type")]
+    [DataRow(CliBool.False, "mssql", DatabaseType.MSSQL, DisplayName = "Init command with '--graphql.nested-create.enabled false' for MsSql database type")]
+    [DataRow(CliBool.None, "mssql", DatabaseType.MSSQL, DisplayName = "Init command without '--graphql.nested-create.enabled true' for MsSql database type")]
```
Scenario : `--graphql.nested-create.enabled` is used without providing `true/false`. 
So, the complete CLI command will look like: `dab init --database-type mssql --graphql.nested-create.enabled`. 


On L142 (and other lines that validate the same scenario), removed `true` after `--graphql.nested-create.enabled`

severussundar commented on schemas/dab.draft.schema.json:
```diff
@@ -175,6 +175,25 @@
+            },
+            "nested-mutations": {
+              "type": "object",
+              "description": "Feature flags for nested mutation operations",
```
Updated the description to `Configuration properties for nested mutation operations`

severussundar commented:
/azp run

severussundar commented:
Rename of the CLI options, JSON fields etc. (replacing nested mutations/nested create) with a different name 
 will be taken up in a follow-up PR. Created [this](https://github.com/Azure/data-api-builder/issues/2090) issue for tracking the same.

Aniruddh25 reviewed (approved):
LGTM! Thanks for your patience in the review.

seantleonard reviewed (approved):
Lgtm! thanks for addressing the various iterations of feedback.
